---
layout: post
title:  "Microbiology from the Command Line"
author: at
categories: [ Jekyll, tutorial ]
image: assets/images/breakfast.jpg
featured: true
hidden: true
---

This website collects some notes, tutorials and handsout used for
*microbial bioinformatics training* events.

#### Introduction to the Bash scripting

You learnt how to work from the command line, and now want to start automating
your actions with scripts. This is a gentle introduction for you.

* [Introduction and first examples]({{ site.baseurl }}{% link _posts/2019-01-02-Bash-tutorial-2.md %})
* [For loops]({{ site.baseurl }}{% link _posts/2019-01-02-Bash-tutorial-2.md %})
* [`if` conditionals]({{ site.baseurl }}{% link _posts/2019-01-03-Bash-tutorial-3.md %})
* [Safety net]({{ site.baseurl }}{% link _posts/2019-01-04-Bash-tutorial-4.md %})
* [Getting parameters]({{ site.baseurl }}{% link _posts/2019-01-05-Bash-tutorial-5.md %})

#### Metabarcoding workshop

* Day1 - [Introduction, Qiime2 and QC]({% post_url 2021-02-01-Metabarcoding-1.md %})
* Day2 - [From the feature table to diversity analyses]({% post_url 2021-02-02-Metabarcoding-2.md %})
* Day3 - [Analyse your data in R]({% post_url 2021-02-03-Metabarcoding-3.md %})

#### Metagenomics workshop: classifying your reads

* Day1 - [Introduction and preliminary aspects]({% post_url 2021-03-01-Metagenomics-classification-1.md %})
* Day2 - [Kraken2, Bracken, Krona and MultiQC]({% post_url 2021-03-02-Metagenomics-classification-2.md %})
* Day3 - [Analyse your data in R]({% post_url 2021-03-03-Metagenomics-classification-3.md %})
