---
layout: post
title:  "Taxonomic profiling of whole metagenome shotgun (day 1)"
author: at
categories: [ metagenomics, tutorial ]
image: assets/images/xray-bact.jpg
---

> On our first day we well cover the concepts behind taxonomic classification using Kraken2 (and Bracken), and see how to remove host reads and perform the quality checks (and filtering).

## Our dataset

We have a set of samples from an ongoing study gut metagenomics by 
Aimee Parker (Quadram Institute) and co-workers. Part of her study
has been [described in a pre-print](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3811833), we took some samples to practice our classification
skills.

 * Sample\_3
 * Sample\_6
 * Sample\_30
 * Sample\_4
 * Sample\_22
 * Sample\_25
 * Sample\_13
 * Sample\_31
 

## Initial QC

A well known tool for initial QC of FASTQ reads is FastQC.

```
mkdir fastqc-reports
fastqc --outdir fastqc-reports --threads 4 reads/*.fastq.gz
```

## Quality filtering

[fastp](https://github.com/OpenGene/fastp#readme) was designed
to have "good defaults", but it's always a good idea to check
what our options are to tweak the parameters.

```
for i in reads/*R1*gz;
do
  fastp  -w 16 -i $i -I ${i/R1/R1} \
   -o filt/$(basename $i) -O filt/$(basename  ${i/_R1/_R2}) \
   --detect_adapter_for_pe \
   --length_required 100 \
   --overrepresentation_analysis;
 done
```

## Host removal

## Reads filtering