---
layout: post
title:  "Metabarcoding workshop (day 3)"
author: ra
categories: [ metabarcoding, 16S, tutorial ]
image: assets/images/green-bact.jpg
---

## From the Command line to R

In your _RStudio_ you will need two libraries:

* PhyloSeq
* ggplot2

### Installing packages

:bulb: If you  have _PhyloSeq_ installed, skip this section.

The repository for bioinformatics libraries for R is called 
[Bioconductor](https://bioconductor.org/install/) 
(we can think of it as the _bioconda_ channel for the _Miniconda_ package manager.). 
To install the Bioconductor package manager we can run this command:
```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.13")
```

To install one or more (bioconductor) package:
```r
BiocManager::install( c("phyloseq", "microbiome") )
```

## Phyloseq object

[Phyloseq](https://joey711.github.io/phyloseq/) is a great package that brings to life a single object that integrates:
* Metadata (mapping file)
* A feature table (OTU table)
* The taxonomy assignments
* The feature tree

A single PhyloSeq object contains our whole metabarcoding experiment!

As different tools will generate different files, a great source of information is
the [importing phyloseq data tutorial](https://joey711.github.io/phyloseq/import-data.html).

In our case we have two options:
* Importing data from the Qiime 2 artifacts
* Using the PhyloSeq object automatically generated by Dadaist2

## A preliminary exploration of the phyloseq object

Once that we have our PhyloSeq object, we can use the many functions that [PhyloSeq](https://joey711.github.io/phyloseq/) 
(and other PhyloSeq-based libraries, like [microbiome](https://microbiome.github.io/tutorials/)) provide.

Here we provide an example based on the dataset analysed in the workshop ("MiSeq SOP"), available 
 as an **[R markdown]({{ site.baseurl }}/rmd/phyloseq.html)** script made by Rebecca Ansorge.

:bulb: If you need the phyloseq object, you can [download it from here]({{ '/data/phyloseq.rds' | prepend: site.baseurl }}).