---
layout: post
title:  "Kaiju"
author: at
categories: [ metagenomics, tutorial ]
hidden: false
---

[Kaiju](https://github.com/bioinformatics-centre/kaiju) can be installed, as usual, from Conda.
Like Kraken2, we have access to [pre-built databases](http://kaiju.binf.ku.dk/server), and for this
tutorial we used the _nr 2021-02-24_ (52 GB).

```
kaiju -t $DB/kaiju/nodes.dmp -f $DB/kaiju/kaiju_db_nr.fmi  -o kaiju.tsv -z 32 -v  -i subsampled_R1.fq.gz-j subsampled_R2.fq.gz 
```

### Generate a report

Kaiju won't generate a report on-the-fly, but ships a program to do one (that can be automatically imported
by [MultiQC]({{ site.baseurl}}{% link _posts/2021-03-06-MultiQC.md %})).

```
kaiju2table -t /data/db/kaiju/nodes.dmp -n /data/db/kaiju/names.dmp -r phylum -o kaiju-phylum.tsv kaiju.tsv 
kaiju2table -t /data/db/kaiju/nodes.dmp -n /data/db/kaiju/names.dmp -r species -o kaiju-species.tsv kaiju.tsv 
```

### Exporting to Krona

Kaiju also ships a small utility to prepare a tabular file to be imported in [Krona]({{site.baseurl}}{% link _posts/2021-03-06-Kraken-to-Krona.md %}).
If we want the unclassified to be reported, we need to add the `-u` flag.

```
kaiju2krona -t /data/db/kaiju/nodes.dmp -n /data/db/kaiju/names.dmp -i kaiju.tsv -o kaiju.krona  -u
ktImportText -o kaiju.out.html kaiju.krona
```