---
layout: post
title:  "Using MultiQC to plot Kraken2 data"
author: at
categories: [ metagenomics, tutorial ]
hidden: true
---

### Installing MultiQC

We installed _MultiQC_ via conda, but in case we need to install it again:

```
conda install -c bioconda multiqc
```

### What is MultiQC

[MultiQC](https://multiqc.info) is a fantastic tool that can aggregate outputs from different bioinformatics
programs in a single report.
MultiQC is capable of understanding the output of a hunder tools (including: _fastp_, _cutadapt_, _prokka_, **kaiju**, _quast_...)

:mag: to see the flexibility of MultiQC, you can see [an example complete report](https://multiqc.info/examples/ngi-rna/Test_NGI_Project_multiqc_report.html).


 We will combine our _fastp_ and _Kraken2_ classifications
to have a single report.

It works with a bit of magic: it scans all your files to check if some looks like a bioinformatic output. Sometimes the filename is important as well, for example 
for _fastp_ it will use the `.json` file, that should be in the `*.fastp.json` format.

```
cd ~/kraken-ws
for i in reports/*.json;
do
  mv $i ${i/json/fastp.json}
done
```

We can first create a report just based on FASTP with: 

```
multiqc -o fastp-report reports/
```

With *-o* we specify the output directory, then we need to tell where MultiQC should scan for known files.
The output should be similar to [this one](https://telatin.github.io/microbiome-bioinformatics/data/multiqc/fastp-report/).
If you check your report, you will notice that MultiQC thinks our samples are called _Samplename\_1_, because 
it's taken from the first pair. 

We want to remove the `_1` to make it mergeable with Kraken:

```
sed -i 's/_1//' reports/*.json
```

Now we can combine _fastp_ and _Kraken2_:
```
multiqc -o multiqc reports/ kraken/
```

:mag: The output should be [like this one](https://telatin.github.io/microbiome-bioinformatics/data/multiqc/)

### Citation

Philip Ewels, Måns Magnusson, Sverker Lundin and Max Käller, [MultiQC: Summarize analysis results for multiple tools and samples in a single report](https://doi.org/10.1093/bioinformatics/btw354), Bioinformatics (2016)